Pod crash on panic of interface conversation

I've some pods that sometimes crash with the following panic.
This happens from time to time:

```
{"level":"info","time":1695111317,"message":"Starting server listening on :9100"}
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 10 [running]:
main.getMetrics()
	/code/main.go:112 +0x930
created by main.main
	/code/main.go:148 +0x76
```

Maybe worth validating the conversation of interfaces, and if it fails to be what the code expects it to log it as error but keep running?

--- a/Dockerfile
+++ b/Dockerfile
@@ -12,7 +12,7 @@ COPY . .
 
 RUN go mod download
 
-RUN CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -a -o app main.go
+RUN CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -ldflags="-X main.Commit=$(git rev-parse HEAD)" -a -o app main.go
 
 FROM --platform=${BUILDPLATFORM:-linux/amd64} gcr.io/distroless/static:nonroot
 WORKDIR /

--- a/main.go
+++ b/main.go
@@ -17,6 +17,7 @@ import (
 	"os"
 	"path/filepath"
 	"runtime"
+	"runtime/debug"
 	"strconv"
 	"time"
 )
@@ -27,6 +28,18 @@ var (
 	currentNode    string
 	sampleInterval int
 )
+
+var Commit = func() string {
+	if info, ok := debug.ReadBuildInfo(); ok {
+		for _, setting := range info.Settings {
+			if setting.Key == "vcs.revision" {
+				return setting.Value
+			}
+		}
+	}
+
+	return ""
+}()
 
 func getEnv(key, fallback string) string {
 	if value, ok := os.LookupEnv(key); ok {
@@ -123,7 +136,6 @@ func getMetrics() {
 			os.Exit(1)
 		}
 		log.Debug().Msg(fmt.Sprintf("Fetched proxy stats from node : %s", currentNode))
-		//var raw map[string]interface{}
 		var data ephemeralStorageMetrics
 		_ = json.Unmarshal(content, &data)
 
@@ -173,11 +185,10 @@ func main() {
 	go getMetrics()
 	port := getEnv("METRICS_PORT", "9100")
 	http.Handle("/metrics", promhttp.Handler())
-	log.Info().Msg(fmt.Sprintf("Starting server listening on :%s", port))
+	log.Info().Msg(fmt.Sprintf("Starting server listening on :%s (version=%s)", port, Commit))
 	err := http.ListenAndServe(fmt.Sprintf(":%s", port), nil)
 	if err != nil {
 		log.Error().Msg(fmt.Sprintf("Listener Failed : %s\n", err.Error()))
 		panic(err.Error())
 	}
-
 }

--- a/main.go
+++ b/main.go
@@ -17,7 +17,6 @@ import (
 	"os"
 	"path/filepath"
 	"runtime"
-	"runtime/debug"
 	"strconv"
 	"time"
 )
@@ -27,19 +26,9 @@ var (
 	clientset      *kubernetes.Clientset
 	currentNode    string
 	sampleInterval int
+
+	Commit string
 )
-
-var Commit = func() string {
-	if info, ok := debug.ReadBuildInfo(); ok {
-		for _, setting := range info.Settings {
-			if setting.Key == "vcs.revision" {
-				return setting.Value
-			}
-		}
-	}
-
-	return ""
-}()
 
 func getEnv(key, fallback string) string {
 	if value, ok := os.LookupEnv(key); ok {

--- a/main.go
+++ b/main.go
@@ -17,12 +17,16 @@ import (
 	"os"
 	"path/filepath"
 	"runtime"
+	"strconv"
 	"time"
 )
 
-var inCluster string
-var clientset *kubernetes.Clientset
-var currentNode string
+var (
+	inCluster      string
+	clientset      *kubernetes.Clientset
+	currentNode    string
+	sampleInterval int
+)
 
 func getEnv(key, fallback string) string {
 	if value, ok := os.LookupEnv(key); ok {
@@ -74,6 +78,23 @@ func getK8sClient() {
 	}
 }
 
+type ephemeralStorageMetrics struct {
+	Node struct {
+		NodeName string `json:"nodeName"`
+	}
+	Pods []struct {
+		PodRef struct {
+			Name      string `json:"name"`
+			Namespace string `json:"namespace"`
+		}
+		EphemeralStorage struct {
+			AvailableBytes float64 `json:"availableBytes"`
+			CapacityBytes  float64 `json:"capacityBytes"`
+			UsedBytes      float64 `json:"usedBytes"`
+		} `json:"ephemeral-storage"`
+	}
+}
+
 func getMetrics() {
 
 	opsQueued := prometheus.NewGaugeVec(prometheus.GaugeOpts{
@@ -94,6 +115,7 @@ func getMetrics() {
 
 	log.Debug().Msg(fmt.Sprintf("getMetrics has been invoked"))
 	currentNode = getEnv("CURRENT_NODE_NAME", "")
+	sampleInterval, _ = strconv.Atoi(getEnv("SCRAPE_INTERVAL", "15"))
 	for {
 		content, err := clientset.RESTClient().Get().AbsPath(fmt.Sprintf("/api/v1/nodes/%s/proxy/stats/summary", currentNode)).DoRaw(context.Background())
 		if err != nil {
@@ -101,22 +123,25 @@ func getMetrics() {
 			os.Exit(1)
 		}
 		log.Debug().Msg(fmt.Sprintf("Fetched proxy stats from node : %s", currentNode))
-		var raw map[string]interface{}
-		_ = json.Unmarshal(content, &raw)
+		//var raw map[string]interface{}
+		var data ephemeralStorageMetrics
+		_ = json.Unmarshal(content, &data)
 
-		nodeName := raw["node"].(map[string]interface{})["nodeName"].(string)
-		for _, element := range raw["pods"].([]interface{}) {
-			ref := element.(map[string]interface{})["podRef"].(map[string]interface{})
-			podName := ref["name"].(string)
-			podNamespace := ref["namespace"].(string)
-			usedBytes := element.(map[string]interface{})["ephemeral-storage"].(map[string]interface{})["usedBytes"].(float64)
-
+		nodeName := data.Node.NodeName
+		for _, pod := range data.Pods {
+			podName := pod.PodRef.Name
+			podNamespace := pod.PodRef.Namespace
+			usedBytes := pod.EphemeralStorage.UsedBytes
+			if podNamespace == "" || (usedBytes == 0 && pod.EphemeralStorage.AvailableBytes == 0 && pod.EphemeralStorage.CapacityBytes == 0) {
+				log.Warn().Msg(fmt.Sprintf("pod %s/%s on %s has no metrics on its ephemeral storage usage", podName, podNamespace, nodeName))
+				log.Warn().Msg(fmt.Sprintf("raw content %v", content))
+			}
 			opsQueued.With(prometheus.Labels{"pod_namespace": podNamespace, "pod_name": podName, "node_name": nodeName}).Set(usedBytes)
 
 			log.Debug().Msg(fmt.Sprintf("pod %s/%s on %s with usedBytes: %f", podNamespace, podName, nodeName, usedBytes))
 		}
 
-		time.Sleep(15 * time.Second)
+		time.Sleep(time.Duration(sampleInterval) * time.Second)
 	}
 }
 

